### ISSUE TITLE
Add a language filter to preferences

### ISSUE THREAD

**Issue by asaunier** [c2corg/v6_ui#1382]
In the feed preferences page https://www.camptocamp.org/preferences

See the related API issue at https://github.com/c2corg/v6_api/issues/608

**Comment by desnoes** [comment]
For information, I put here a description of how it worked in V5.

Sur la V5, y compris pour les invités, la langue de l'interface était présélectionnée par défaut en fonction de la langue du navigateur. On pouvait changer ensuite bien sûr.
La personnalisation de la langue des sorties/forum par défaut (y compris pour les invités) dépendaient de cette langue d'interface :

Si la langue de l'interface est FR, EN, DE ou EU : pas de personnalisation de langue, tout est affiché.
Si la langue de l'interface est ES ou CA : personnalisation en ES, CA, EN.
Si la langue de l'interface est IT : personnalisation en IT.
Ceci afin d'éliminer le FR dans les langues minoritaires mais un peu actives, tout en évitant d'avoir un contenu qui ne bouge pas pour les langues très peu actives.

L'algo a changé plusieurs fois en fonction du nb de sorties par mois dans chaque langue.
Il se trouve là : https://github.com/c2corg/camptocamp.org/blob/master/apps/frontend/lib/c2cPersonalization.class.php#L30-L73

**Comment by asaunier** [comment]
Tag "high prio" and milestone "6.3" are not very consistent :P

**Comment by asaunier** [comment]
@desnoes I have the feeling a language filter is a bit a non sense, bringing extra work just to remove content that could interest people.

In the home page, the forum block will anyway not be managed by preferences filters, but by the prefs set directly in the forum.

In the activity feed, you can already filter on regions (and activities), which already strips out most of the content that don't interest you.

True, you might still get reports written in other languages for outings done in your selected areas. But why hidding them? It's still useful data about the conditions, etc. : 
* Pictures are not language-dependant and are definitely interesting.
* Figures data (overall condition ratings, conditions per level - especially the amount of snow at various elevations - , crowding, hut status, elevation of the snow, etc.) are not or not much language-dependent.

Of course the text comments might be difficult to read if you don't understand the language but browsers often detect "foreign language" contents and propose a translation. I guess it will generalize progressively. Or we can add a automatic translating service (as in v5) with a button in documents whose language does not match the current interface language.

To summarize, I strongly recommend not to implement a language filter but rather to make sure "foreign" contents can be autotranslated on demand.

**Comment by desnoes** [comment]
@asaunier: this features was a strong demand from italian contributors (cf. https://forum.camptocamp.org/t/internationalisation-et-v6/176634 and https://forum.camptocamp.org/t/preselection-de-la-langue-dans-les-entrees-du-fil-dactivite/177316) and was tagged as a priority but the CA. I am not sure the quality of an automatic translation is efficient enough for the home page of the site. It is sufficient if you really want to have information about a route or conditions (key words are enough), but it is not performant if you want to pleasant reading. If you really want to reopen this discussion, this should be discuss in the forum.

**Comment by asaunier** [comment]
Specs for the UI part:
* [ ] Add a languages selector (multi-choice) to the "Preferences" page https://www.camptocamp.org/preferences similar to the activities selector.
* [ ] By default no language is selected => no lang filtering (like for the activities).
* [ ] Selecting or deselecting a lang triggers a request to the API preferences service (https://api.camptocamp.org/users/preferences) with the new set of filters (activities, areas, langs). TODO: add langs to the current API request done when updating the prefs.
* [ ] When loading the prefs page, the langs prefs are retrieved from the API preferences service along with the actrivities and areas prefs => TODO: get langs prefs from API and set the langs selector accordingly.

**Comment by desnoes** [comment]
I observed a bug only on demo. 
Selected language : german
Activity: mountain bike
here is the result (KO)

![bug fil activite](https://user-images.githubusercontent.com/19892089/30925706-0c9806c4-a3b3-11e7-9b3f-5a54e729182a.png)
It seems the feed cannot return 'no result'.

**Comment by desnoes** [comment]
Preferences set: MTB and german 
creation of a route (MTB and german) => it is shown in the feed (OK).
creation of an article (MTB and italian) => it is not shown in the feed (OK)
Translation of this article in german => it is shown in the feed but with the wrong language (italian) (KO)

![bug fil activite 2](https://user-images.githubusercontent.com/19892089/30926729-68d5daa8-a3b6-11e7-8ee3-1c99b547fcc4.png)

**Comment by asaunier** [comment]
@desnoes Thanks for testing!

Documents are referenced in the feed as a single entry (for a given user) containing the following attributes:
* ``activities``: list of activities => used for the "activities" filter
* ``langs``: list of available lang versions of the document => used for the "langs" filter
* ``area_ids``: list of associated regions => used for the "regions" filter
* ``user_ids``: list of related users (document creator + associated participants for outings) => used to filter on users you follow
* ``user_id``: id of the user who made the action: created the document, uploaded photos, associated participants (for outings). Distinct cards are used for the same document when distinct users perform actions on a given document (eg. if I create an outing and if you upload images to the outing, there will be one card for my actions and one for yours).

Document editions are not referenced in the feed. Translating a document (new lang version) is considered as a document editing, I think => adding or deleting a lang version of the document only updates the ``langs`` attribute but does not make the document go up in the feed (if you translate a document created a long time ago, the card won't be displayed at the top of the feed as when you create a new document).

> Translation of this article in german => it is shown in the feed (OK) but with the wrong language (italian) (KO)

This behaviour is consistent:
* the ``langs`` attribute was first ``it`` when you created the article. Then it was updated to ``it,de`` when you created the german version.
* the card now appears in the feed because it matches your langs filter
* the lang version shown in the card does not rely in the langs feed filter settings (it's just a list of languages, not even ordered). It is based on the usual languages priority configuration which is 
> use the interface language if available, if not use the first lang version available using the priority defined in https://github.com/c2corg/v6_common/blob/master/c2corg_common/attributes.py#L3

In your example the italian version is served because:
* no version is available in your interface language (``fr``)
* ``it`` comes before ``de`` in https://github.com/c2corg/v6_common/blob/master/c2corg_common/attributes.py#L3

> Outing (muti-activities incluting MTB, first version of the outing in french and then translated in german) => it is shown in the feed (OK) but with the wrong language (french) (KO)

Same than above: french is your interface language, since the document is available in french, the card is served in french. That's logical and again is not related to the langs filter.

By the way this logic is already used in production for the feed. No change was done about it. 

> Then i deleted the german version of the outing = > the french version is still shown in the feed (KO)

Mmmm this is indeed not normal: ``de`` should have been removed from the ``langs`` attribute of the card, thus should no longer match your lang filter. I will check if something is missing.

**Comment by asaunier** [comment]
@desnoes PR https://github.com/c2corg/v6_api/pull/693 should fix 
> Then i deleted the german version of the outing = > the french version is still shown in the feed (KO)

Concerning 
> Selected language : german
Activity: mountain bike
here is the result (KO)
It seems the feed cannot return 'no result'.

The feed can actually return "no result". In my instance with filters "basque" + "MTB" + "Valais E", I actually get no result (by the way the displayed message seems to be not the good one :P):
![capture du 2017-09-28 12-23-22](https://user-images.githubusercontent.com/1192331/30961920-222773d6-a448-11e7-86e8-07485fec1c33.png)

This is rather related to the weird behaviours sometimes observed on the demo and described at https://github.com/c2corg/v6_api/pull/690#issuecomment-332511232

As for now I have no explanation.

**Comment by asaunier** [comment]
> As for now I have no explanation.

I think I have!

It seems to me that all cards not matching the preferences filters are related to followed users! That's an expected behaviour.

**Comment by desnoes** [comment]
@asaunier : many thanks for your explanations !
